jax/_src/numpy/lax_numpy.py

```diff
@@ -95,24 +95,26 @@
 printoptions = np.printoptions
 set_printoptions = np.set_printoptions
 
-# We want isinstance(x, np.ndarray) checks in user code to work with the our
-# array-like types, including DeviceArray and UnshapedArray (i.e. the abstract
-# array base class). We can override the isinstance behavior directly, without
-# having the complexity of multiple inheritance on those classes, by defining
-# the ndarray class to have a metaclass with special __instancecheck__ behavior.
-_arraylike_types = (np.ndarray, UnshapedArray, DeviceArray)
+# ndarray is defined as an virtual abstract base class.
 
-class _ArrayMeta(type(np.ndarray)):  # type: ignore
+class ArrayMeta(abc.ABCMeta):
   """Metaclass for overriding ndarray isinstance checks."""
 
   def __instancecheck__(self, instance):
+    # Allow tracer instances that avals that are instances of UnshapedArray.
+    # We could instead just declare Tracer an instance of the ndarray type, but
+    # there can be traced values that are not arrays. The main downside here is
+    # that isinstance(x, ndarray) might return true but
+    # issubclass(type(x), ndarray) might return false for an array tracer.
```

Comment:
Should we also overload `__subclasscheck__` to address this concern?

Reply:
I don't think it can be done.  `__subclasscheck__` receives a type, not a value. Whether a tracer is an array or not is a property of a value, not a type. We could rectify this by defining a new `ArrayTracer` subclass of `Tracer`, perhaps.